In [1]:
import pandas as  pd 

In [4]:
df=pd.read_csv("heart.csv")
df

Age Sex ChestPainType  RestingBP  Cholesterol  FastingBS RestingECG  \
0     40   M           ATA        140          289          0     Normal   
1     49   F           NAP        160          180          0     Normal   
2     37   M           ATA        130          283          0         ST   
3     48   F           ASY        138          214          0     Normal   
4     54   M           NAP        150          195          0     Normal   
..   ...  ..           ...        ...          ...        ...        ...   
913   45   M            TA        110          264          0     Normal   
914   68   M           ASY        144          193          1     Normal   
915   57   M           ASY        130          131          0     Normal   
916   57   F           ATA        130          236          0        LVH   
917   38   M           NAP        138          175          0     Normal   

     MaxHR ExerciseAngina  Oldpeak ST_Slope  HeartDisease  
0      172              N      0.0       Up             0  
1      156              N      1.0     Flat             1  
2       98              N      0.0       Up             0  
3      108              Y      1.5     Flat             1  
4      122              N      0.0       Up             0  
..     ...            ...      ...      ...           ...  
913    132              N      1.2     Flat             1  
914    141              N      3.4     Flat             1  
915    115              Y      1.2     Flat             1  
916    174              N      0.0     Flat             1  
917    173              N      0.0       Up             0  

[918 rows x 12 columns]

In [5]:
col=df.select_dtypes(include="object").columns
df=pd.get_dummies(df,columns=col)
df

Age  RestingBP  Cholesterol  FastingBS  MaxHR  Oldpeak  HeartDisease  \
0     40        140          289          0    172      0.0             0   
1     49        160          180          0    156      1.0             1   
2     37        130          283          0     98      0.0             0   
3     48        138          214          0    108      1.5             1   
4     54        150          195          0    122      0.0             0   
..   ...        ...          ...        ...    ...      ...           ...   
913   45        110          264          0    132      1.2             1   
914   68        144          193          1    141      3.4             1   
915   57        130          131          0    115      1.2             1   
916   57        130          236          0    174      0.0             1   
917   38        138          175          0    173      0.0             0   

     Sex_F  Sex_M  ChestPainType_ASY  ...  ChestPainType_NAP  \
0        0      1                  0  ...                  0   
1        1      0                  0  ...                  1   
2        0      1                  0  ...                  0   
3        1      0                  1  ...                  0   
4        0      1                  0  ...                  1   
..     ...    ...                ...  ...                ...   
913      0      1                  0  ...                  0   
914      0      1                  1  ...                  0   
915      0      1                  1  ...                  0   
916      1      0                  0  ...                  0   
917      0      1                  0  ...                  1   

     ChestPainType_TA  RestingECG_LVH  RestingECG_Normal  RestingECG_ST  \
0                   0               0                  1              0   
1                   0               0                  1              0   
2                   0               0                  0              1   
3                   0               0                  1              0   
4                   0               0                  1              0   
..                ...             ...                ...            ...   
913                 1               0                  1              0   
914                 0               0                  1              0   
915                 0               0                  1              0   
916                 0               1                  0              0   
917                 0               0                  1              0   

     ExerciseAngina_N  ExerciseAngina_Y  ST_Slope_Down  ST_Slope_Flat  \
0                   1                 0              0              0   
1                   1                 0              0              1   
2                   1                 0              0              0   
3                   0                 1              0              1   
4                   1                 0              0              0   
..                ...               ...            ...            ...   
913                 1                 0              0              1   
914                 1                 0              0              1   
915                 0                 1              0              1   
916                 1                 0              0              1   
917                 1                 0              0              0   

     ST_Slope_Up  
0              1  
1              0  
2              1  
3              0  
4              1  
..           ...  
913            0  
914            0  
915            0  
916            0  
917            1  

[918 rows x 21 columns]

In [6]:
x=df.drop(["HeartDisease"],axis=1)
x

Age  RestingBP  Cholesterol  FastingBS  MaxHR  Oldpeak  Sex_F  Sex_M  \
0     40        140          289          0    172      0.0      0      1   
1     49        160          180          0    156      1.0      1      0   
2     37        130          283          0     98      0.0      0      1   
3     48        138          214          0    108      1.5      1      0   
4     54        150          195          0    122      0.0      0      1   
..   ...        ...          ...        ...    ...      ...    ...    ...   
913   45        110          264          0    132      1.2      0      1   
914   68        144          193          1    141      3.4      0      1   
915   57        130          131          0    115      1.2      0      1   
916   57        130          236          0    174      0.0      1      0   
917   38        138          175          0    173      0.0      0      1   

     ChestPainType_ASY  ChestPainType_ATA  ChestPainType_NAP  \
0                    0                  1                  0   
1                    0                  0                  1   
2                    0                  1                  0   
3                    1                  0                  0   
4                    0                  0                  1   
..                 ...                ...                ...   
913                  0                  0                  0   
914                  1                  0                  0   
915                  1                  0                  0   
916                  0                  1                  0   
917                  0                  0                  1   

     ChestPainType_TA  RestingECG_LVH  RestingECG_Normal  RestingECG_ST  \
0                   0               0                  1              0   
1                   0               0                  1              0   
2                   0               0                  0              1   
3                   0               0                  1              0   
4                   0               0                  1              0   
..                ...             ...                ...            ...   
913                 1               0                  1              0   
914                 0               0                  1              0   
915                 0               0                  1              0   
916                 0               1                  0              0   
917                 0               0                  1              0   

     ExerciseAngina_N  ExerciseAngina_Y  ST_Slope_Down  ST_Slope_Flat  \
0                   1                 0              0              0   
1                   1                 0              0              1   
2                   1                 0              0              0   
3                   0                 1              0              1   
4                   1                 0              0              0   
..                ...               ...            ...            ...   
913                 1                 0              0              1   
914                 1                 0              0              1   
915                 0                 1              0              1   
916                 1                 0              0              1   
917                 1                 0              0              0   

     ST_Slope_Up  
0              1  
1              0  
2              1  
3              0  
4              1  
..           ...  
913            0  
914            0  
915            0  
916            0  
917            1  

[918 rows x 20 columns]

In [7]:
y=df["HeartDisease"]
y

0      0
1      1
2      0
3      1
4      0
      ..
913    1
914    1
915    1
916    1
917    0
Name: HeartDisease, Length: 918, dtype: int64

In [8]:
from sklearn.linear_model import LogisticRegression
from sklearn.ensemble import RandomForestClassifier
from sklearn.svm import SVC
from sklearn.tree import DecisionTreeClassifier

from sklearn.model_selection import train_test_split

In [11]:
x_train,x_test,y_train,y_test=train_test_split(x,y,test_size=0.25)

len(x_train) , len(x_test)

(688, 230)

In [15]:
def get_score(model,xtrain,ytrain,xtest,ytest):
    model.fit(xtrain,ytrain)
    return model , model.score(xtest,ytest)

In [16]:
print(get_score(LogisticRegression(),x_train,y_train,x_test,y_test))
print(get_score(RandomForestClassifier(),x_train,y_train,x_test,y_test))
print(get_score(SVC(),x_train,y_train,x_test,y_test))
print(get_score(DecisionTreeClassifier(),x_train,y_train,x_test,y_test))


(LogisticRegression(), 0.8652173913043478)
(RandomForestClassifier(), 0.9)
(SVC(), 0.7434782608695653)
(DecisionTreeClassifier(), 0.8130434782608695)


C:\Users\Janak\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.10_qbz5n2kfra8p0\LocalCache\local-packages\Python310\site-packages\sklearn\linear_model\_logistic.py:458: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


In [17]:
from sklearn.model_selection import cross_val_score

print(sum(cross_val_score(LogisticRegression(),x,y))/5)
print(sum(cross_val_score(RandomForestClassifier(),x,y))/5)
print(sum(cross_val_score(SVC(),x,y))/5)
print(sum(cross_val_score(DecisionTreeClassifier(),x,y))/5)

C:\Users\Janak\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.10_qbz5n2kfra8p0\LocalCache\local-packages\Python310\site-packages\sklearn\linear_model\_logistic.py:458: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(
C:\Users\Janak\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.10_qbz5n2kfra8p0\LocalCache\local-packages\Python310\site-packages\sklearn\linear_model\_logistic.py:458: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stab

0.8332858161083394
0.8332739368020906
0.6873129009265859
0.7482775005939654


Random forest & Logistic reg is performing well  

In [74]:
from sklearn.model_selection import GridSearchCV

lr=GridSearchCV(LogisticRegression(),{
    'solver':['newton-cg','lbfgs','liblinear','sag','saga']
},cv=5,return_train_score=False)

lr.fit(x,y)
lr.best_score_ ,lr.best_params_


C:\Users\Janak\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.10_qbz5n2kfra8p0\LocalCache\local-packages\Python310\site-packages\sklearn\linear_model\_logistic.py:458: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(
C:\Users\Janak\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.10_qbz5n2kfra8p0\LocalCache\local-packages\Python310\site-packages\sklearn\linear_model\_logistic.py:458: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stab

(0.8332858161083394, {'solver': 'lbfgs'})

In [75]:
rf=GridSearchCV(RandomForestClassifier(),{
    'n_estimators':[1,10,20,30,40,50],
    'max_depth':[20,30,35,40,45,50]
},cv=5,return_train_score=False)

rf.fit(x,y)
rf.best_score_,rf.best_params_


(0.8474340698503207, {'max_depth': 30, 'n_estimators': 50})

**random forest and logistic are performing at same level with score and param of**  

*logistic reg (0.8332858161083394, {'solver': 'lbfgs'})* <br />
*random forest (0.8376098835827988, {'max_depth': 30, 'n_estimators': 40})*


In [78]:
x[1:2] , y[1]
df[1:2]

Age  RestingBP  Cholesterol  FastingBS  MaxHR  Oldpeak  HeartDisease  \
1   49        160          180          0    156      1.0             1   

   Sex_F  Sex_M  ChestPainType_ASY  ...  ChestPainType_NAP  ChestPainType_TA  \
1      1      0                  0  ...                  1                 0   

   RestingECG_LVH  RestingECG_Normal  RestingECG_ST  ExerciseAngina_N  \
1               0                  1              0                 1   

   ExerciseAngina_Y  ST_Slope_Down  ST_Slope_Flat  ST_Slope_Up  
1                 0              0              1            0  

[1 rows x 21 columns]

In [79]:
lr.predict(x[1:2])  ## wrong 

array([0], dtype=int64)

In [80]:
rf.predict(x[1:2])  ## right 

array([1], dtype=int64)

In [81]:
df[2:3]

Age  RestingBP  Cholesterol  FastingBS  MaxHR  Oldpeak  HeartDisease  \
2   37        130          283          0     98      0.0             0   

   Sex_F  Sex_M  ChestPainType_ASY  ...  ChestPainType_NAP  ChestPainType_TA  \
2      0      1                  0  ...                  0                 0   

   RestingECG_LVH  RestingECG_Normal  RestingECG_ST  ExerciseAngina_N  \
2               0                  0              1                 1   

   ExerciseAngina_Y  ST_Slope_Down  ST_Slope_Flat  ST_Slope_Up  
2                 0              0              0            1  

[1 rows x 21 columns]

In [83]:
lr.predict(x[2:3]) , rf.predict(x[2:3])

(array([0], dtype=int64), array([0], dtype=int64))